### PR TITLE
feat(function): Support cast variant to other data types

### LINKED ISSUE #4777: Casting from Variant into Other types.
**Summary**

Table:
```
 CREATE TABLE `json_test` (
  `a` Variant,
)


insert into json_test values('{"aa":1,"aA":2,"Aa":3}');
```

Query:
```
select sum(get(a, 'aa'))  from json_test;
```

Error:
```
AggregateSumFunction does not support type 'Variant' (while in select group by)
```

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support cast variant to other data types

for example:
```sql
mysql> select get_path(parse_json('{"a":12}'), 'a')::uint8;
+------------------------------------------------------+
| cast(get_path(parse_json('{"a":12}'), 'a') as UInt8) |
+------------------------------------------------------+
|                                                   12 |
+------------------------------------------------------+
1 row in set (0.03 sec)
```

## Changelog

- New Feature
- Build/Testing/CI

## Related Issues

Fixes #4777

### CODE REVIEW

File: tests/suites/0_stateless/02_function/02_0002_function_cast.sql [hunk 1]
```diff
@@ -32,3 +32,57 @@ SELECT toDate(toDateTime('2021-03-05 01:00:00')) = toDate('2021-03-05');
...
+SELECT parse_json('"test"')::date16; -- {ErrorCode 1010}
+SELECT parse_json('"test"')::date32; -- {ErrorCode 1010}
+SELECT parse_json('"test"')::datetime32; -- {ErrorCode 1010}
+SELECT parse_json('"test"')::datetime64; -- {ErrorCode 1010}
+SELECT parse_json('[1,2,3]')::array;
+SELECT parse_json(1)::array;
+SELECT parse_json('"ab"')::array;
+SELECT parse_json('{"a":1,"b":2}')::object;
+SELECT parse_json('"abc"')::object; -- {ErrorCode 1010}
+SELECT parse_json('[1,2,3]')::object; -- {ErrorCode 1010}
```
Comment (sundy-li) [inline tests/suites/0_stateless/02_function/02_0002_function_cast.sql:88]: Missing tests to cover the `JSON::Null` cases.

**Review by sundy-li — APPROVED** [review]
